File: python/binary_tree/binary_tree.py
class Leaf():
  def __init__(self, v, left=None, right=None):
    self.v = v
    self.left = left
    self.right = right

class BinaryTree():
  def __init__(self):
    self.head = None

  def create(self, l):
    if not self.head:
      self.head = l
      return

    top = self.head
    while top.v != l.v:
      if l.v > top.v:
        if not top.right:
          top.right = l
        top = top.right
      else: # l.v < top.v
        if not top.left:
          top.left = l
        top = top.left

  def read(self, v):
    if not self.head:
      return None

    top = self.head
    while top != None and top.v != v:
      if v > top.v:
        top = top.right
      elif v < top.v:
        top = top.left
    return top

  def delete(self, v):
    if not self.head:
      return None

    # find node of value v, also store the parent
    last = None
    top = self.head
    while top.v != v:
      last = top
      if v > top.v:
        top = top.right
      elif v < top.v:
        top = top.left
      if not top: # could not find value
        return None
    lastDirRight = False if last == None else (last.right == top)

    if top.right and top.left:
      self.__deleteWhenLeafHasBothChildren(top)
    elif top.right:
      self.__deleteWhenLeafHasOneChild(top.right, last, lastDirRight)
    elif top.left:
      self.__deleteWhenLeafHasOneChild(top.left, last, lastDirRight)
    else:
      self.__deleteWhenLeafHasOneChild(None, last, lastDirRight)

  def __deleteReplaceTop(self, top, last, lastDirRight):
    if not last:
      self.head = top
    else:
      if lastDirRight:
        last.right = top
      else:
        last.left = top

  def __deleteWhenLeafHasBothChildren(self, top):
    # top.right is guaranteed to not be nil
    minParent = top.right
    minAtRight = top.right
    while minAtRight.left != None:
      minParent = minAtRight
      minAtRight = minAtRight.left

    top.v = minAtRight.v
    if minParent == minAtRight:
      top.right = None
    else:
      minParent.left = minAtRight.right

  def printTree(self):
    self.printRecursive(self.head, 0)

  def printRecursive(self, node, depth):
    if not node:
      return
    self.printRecursive(node.right, depth + 1)
    for i in xrange(depth):
      print '-----|',
    print node.v
    self.printRecursive(node.left, depth + 1)

  def makeLinkedList(self):
    grandparent = None
    parent = self.head
    leftChild = None

    while parent:
      leftChild = parent.left

      if leftChild:
        if grandparent:
          grandparent.right = leftChild
        else:
          self.head = leftChild

        parent.left = leftChild.right
        leftChild.right = parent

        parent = leftChild
      else:
        grandparent = parent
        parent = parent.right

  def makeBalanced(self):
    self.makeLinkedList

    leafCount = 0
    current = self.head
    while current:
      leafCount += 1
      current = current.right

    msbOf = leafCount + 1
    msbShifts = 0
    while (msbOf > 1):
      msbOf = msbOf >> 1
      msbShifts += 1
    levelRotations = (1 << msbShifts) - 1
    firstRotations = leafCount - levelRotations
    self.doBalanceRotations(firstRotations)

    while levelRotations > 1:
      levelRotations /= 2
      self.doBalanceRotations(levelRotations)

  def doBalanceRotations(self, rotations):
    grandparent = None
    parent = self.head
    rightChild = parent.right

    while rotations > 0:
      if not rightChild:
        break

      if grandparent:
        grandparent.right = rightChild
      else:
        self.head = rightChild
      parent.right = rightChild.left
      rightChild.left = parent

      grandparent = rightChild
      parent = grandparent.right
      rightChild = parent.right if parent else None

      rotations -= 1

  def isSymmetricWalk(self):
    if not self.head:
      return True

    return self.isSymmetricWalkRecurse(self.head.left, self.head.right)


  def isSymmetricWalkRecurse(self, left, right):
    if not left and not right:
      return True
    if not left or not right:
      return False

    return self.isSymmetricWalkRecurse(left.left, right.right) and self.isSymmetricWalkRecurse(left.right, right.left)

  def clear(self):
    self.head = None

  def isSymmetricCount(self):
    kq = self.isSymmetricCountRecurse(self.head, [0, 0])
    return kq[0] is kq[1]

  def isSymmetricCountRecurse(self, start, collect):
    if not start:
      return collect

    if start.left:
      collect[0] += 1

    if start.right:
      collect[1] += 1

    self.isSymmetricCountRecurse(start.left, collect)
    self.isSymmetricCountRecurse(start.right, collect)

    return collect

  def reverse(self):
    self.reverseRecurse(self.head)

  def reverseRecurse(self, start):
    if not start:
      return

    temp = start.left
    start.left = start.right
    start.right = temp

    self.reverseRecurse(start.left)
    self.reverseRecurse(start.right)




def ap(t, v):
  t.create(Leaf(v))
  t.printTree()
  print 'count symmetric?', bt.isSymmetricCount()
  print 'walk symmetric?', bt.isSymmetricWalk()
  print

bt = BinaryTree()

ap(bt, 10)
ap(bt, 5)
ap(bt, 7)
ap(bt, 2)
ap(bt, 15)
ap(bt, 12)
ap(bt, 18)

bt.reverse()
bt.printTree()

File: ruby/binary_tree/binary_tree.rb
class Leaf
  attr_accessor :value, :left, :right
  def initialize(value, left=nil, right=nil)
    @value = value
    @left = left
    @right = right
  end
end

class BinaryTree

  attr_reader :head

  def initialize
    @head = nil
  end

  def add(v)
    unless @head
      @head = Leaf.new(v)
      return
    end

    current = @head
    while current.value != v
      if v < current.value
        current.left = Leaf.new(v) unless current.left
        current = current.left
      else
        current.right = Leaf.new(v) unless current.right
        current = current.right
      end
    end
  end

  def find(v)
    return nil unless @head

    current = @head
    while current
      if v == current.value
        return current
      elsif v > current.value
        current = current.right
      else
        current = current.left
      end  
    end
    return nil
  end

  def clear
    @head = nil
  end

  def delete(v)
    return false unless @head

    # find the node by value, save the parent node
    # and save a boolean that indicates if we last went right(t) or left(f)
    parent = nil
    current = @head
    lastWentRight = false
    while current.value != v
      parent = current
      if v > current.value
        lastWentRight = true
        current = current.right
      else
        lastWentRight = false
        current = current.left
      end 
      return false unless current
    end

    # 4 possible conditions:
    # if both children
    # if right child
    # if left child
    # else no children
    if current.right && current.left
      return deletionCaseBothChildren(current)
    elsif current.right
      return deletionReplaceTop(parent, current.right, lastWentRight)
    elsif current.left
      return deletionReplaceTop(parent, current.left, lastWentRight)
    else
      return deletionReplaceTop(parent, nil, lastWentRight)
    end
  end

  def to_a
    inorder_to_a(@head)
  end

  def to_a_preorder
    preorder_to_a(@head)
  end

  def to_a_postorder
    postorder_to_a(@head)
  end

  def find_depth(v)
    current = @head
    depth = 0
    while current
      if current.value == v
        return depth
      elsif v < current.value
        current = current.left
      else # v > current.value
        current = current.right
      end
      depth += 1
    end
    nil
  end

  def depth_max
    depth_max_from(@head)
  end

  def depth_count(depth)
    foundAtDepth = 0
    traverseInOrder(@head, 1, depth) { |node, node_depth|
      if node_depth == depth
        foundAtDepth += 1
      end
    }
    return foundAtDepth
  end

  def depth_find(depth)
    setFoundAtDepth = []
    traverseInOrder(@head, 1, depth) { |node, node_depth|
      if node_depth == depth
        setFoundAtDepth << node.value
      end
    }
    return setFoundAtDepth
  end

  def make_linked_list
    return unless @head
    grandparent = nil
    parent = @head
    leftChild = nil

    # starting at parent = root
    while parent
      leftChild = parent.left

      # if there is a left child
      if leftChild

        # this is a "right rotation"
        # move the left child between the grandparent and parent
        # if there is no grandparent, the left child becomes the head
        if grandparent
          grandparent.right = leftChild
        else
          @head = leftChild
        end
        # now the grandparent.right/head is the left child

        # set the old link to left child to the left childs right child
        parent.left = leftChild.right
        # now the left childs right child is the parent
        leftChild.right = parent

        # what used to be down left, is now up
        # the old left child could still have more children so move up
        parent = leftChild
      else
        # if no left child
        # move down right
        grandparent = parent
        parent = parent.right
      end
    end
  end

  def make_balanced
    # first convert to a linked list
    make_linked_list

    # find the length of the linked list
    # don't just check the max depth, we don't need to check any left nodes
    leafCount = 0
    start = @head
    while start
      leafCount += 1
      start = start.right
    end

    # EX
    # leaf_count = 10
    # count_msb_of = 1011
    # msb_shifts = 3
    # rotations_to_half = 7 (1 << 3 = 1000 then 0111 after -1)
    # first_rotation = 3 (10 - 7)
    # then 7/2 > 1, rotate floor(7/2) times
    # then 3/2 > 1, rotate(3/2) times
    # then 1/2 < 1, finished

    count_msb_of = leafCount + 1
    msb_shifts = 0
    while (1 < count_msb_of)
      count_msb_of = (count_msb_of >> 1)
      msb_shifts += 1
    end
    rotationsToHalf = (1 << msb_shifts) - 1


    firstRotation = leafCount - rotationsToHalf
    balance_rotation(firstRotation)

    while (rotationsToHalf > 1)
      rotationsToHalf /= 2
      balance_rotation(rotationsToHalf)
    end

  end

  def doPrint()
    traverseInOrder(@head) { |n, d|
      (d - 1).times { print "---|" }
      puts "#{n.value}"
    }
  end

  private

  def inorder_to_a(leaf, ret=[])
    return unless leaf
    inorder_to_a(leaf.left, ret)
    ret << leaf.value
    inorder_to_a(leaf.right, ret)
    ret
  end

  def preorder_to_a(leaf, ret=[])
    return unless leaf
    ret << leaf.value
    preorder_to_a(leaf.left, ret)
    preorder_to_a(leaf.right, ret)
    ret
  end

  def postorder_to_a(leaf, ret=[])
    return unless leaf
    postorder_to_a(leaf.left, ret)
    postorder_to_a(leaf.right, ret)
    ret << leaf.value
    ret
  end

  def deletionCaseBothChildren(current)
    return false unless current.right

    # find the node larger than current
    # we expect current.right to not be nil
    minParent = current.right
    minAtRight = current.right
    while minAtRight.left != nil
      minParent = minAtRight
      minAtRight = minAtRight.left
    end

    # swap values
    current.value = minAtRight.value

    # if replacement is immediately to right,
    if minParent == minAtRight
      current.right = nil
    else
      minParent.left = minAtRight.right
    end
    true
  end

  def deletionReplaceTop(parent, successor, directionRight)
    unless parent
      @head = successor
    else
      if directionRight
        parent.right = successor
      else
        parent.left = successor
      end
    end
    true
  end

  def depth_max_from(from=nil, mem=0)
    return mem unless from
    return [
      mem + 1,
      depth_max_from(from.left, mem + 1),
      depth_max_from(from.right, mem + 1)
    ].max
  end

  def traverseInOrder(start, current_depth=1, max_depth=nil)
    return unless start
    return if max_depth && current_depth > max_depth
    traverseInOrder(start.left, current_depth + 1, max_depth) { |e, n|
      yield(e, n)
    }
    yield(start, current_depth)
    traverseInOrder(start.right, current_depth + 1, max_depth) { |e, n|
      yield(e, n)
    }
  end

  # this is a "left rotation"
  # move the parent down left of the right child 
  def balance_rotation(rotations)
    grandparent = nil
    parent = @head
    right_child = @head.right

    # repeat "rotations" # of times
    while rotations > 0
      # in case the # of rotations was too much
      break unless right_child

      # rotate left
      # move the right child of the parent to the right of the grandparent
      # (if no grandparent, it becomes the head)
      if grandparent
        grandparent.right = right_child
      else
        @head = right_child
      end
      # the parents right is now the right childs left child
      parent.right = right_child.left
      # set the right childs left to the parent
      right_child.left = parent
      # end rotate left

      # the new grandparent is the right child
      grandparent = right_child
      # the new parent is to the right of the grandparent
      # which was the old right child (which may not have another right child)
      parent = grandparent.right
      # if the old right child had another right child, we use that
      # for the next iteration
      right_child = parent ? parent.right : nil

      # that was one rotation
      rotations -= 1
    end
  end
end